csi/node_server.go

```diff
@@ -628,19 +629,20 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 	// optionally try to retrieve the volume and check if it's an RWX volume
 	// if it is we let the share-manager clean up the crypto device
 	volume, _ := ns.apiClient.Volume.ById(volumeID)
-	if volume == nil || types.IsDataEngineV1(longhorn.DataEngineType(volume.DataEngine)) {
-		// Currently, only "RWO v1 volumes" and "block device with v1 volume.Migratable is true" supports encryption.
-		sharedAccess := requiresSharedAccess(volume, nil)
-		cleanupCryptoDevice := !sharedAccess || (sharedAccess && volume.Migratable)
-		if cleanupCryptoDevice {
-			cryptoDevice := crypto.VolumeMapper(volumeID)
-			if isOpen, err := crypto.IsDeviceOpen(cryptoDevice); err != nil {
+	dataEngine := string(longhorn.DataEngineTypeV1)
+	if volume != nil {
+		dataEngine = volume.DataEngine
+	}
+	sharedAccess := requiresSharedAccess(volume, nil)
+	cleanupCryptoDevice := !sharedAccess || (sharedAccess && volume.Migratable)
+	if cleanupCryptoDevice {
+		cryptoDevice := crypto.VolumeMapper(volumeID, dataEngine)
+		if isOpen, err := crypto.IsDeviceOpen(cryptoDevice); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		} else if isOpen {
+			log.Infof("Volume %s closing active crypto device %s", volumeID, cryptoDevice)
+			if err := crypto.CloseVolume(volumeID, volume.DataEngine); err != nil {
```

Comment:
It doesn't handle `volume == nil` case

Reply:
Lines 636 to 648, are originally processed when `volume == nil || types.IsDataEngineV1(...)`
The only place using the `volume` is the method `requiresSharedAccess`, and `requiresSharedAccess` will handle whether the volume is nil.

Reply:
Should the func directly use variable `dataEngine` here?

Reply:
>Should the func directly use variable dataEngine here?

Yes, it should use dataEngine. Thanks.

```diff
-if err := crypto.CloseVolume(volumeID, volume.DataEngine); err != nil {
+if err := crypto.CloseVolume(volumeID, dataEngine); err != nil {
```

Reply:
Modified. Thanks.